KEP 2079 - Add PortRange KEP (#2090)
kubernetes/enhancements: keps/sig-network/2079-network-policy-port-range/README.md

undesired traffic.

There are a number of user issues like [kubernetes #67526](https://github.com/kubernetes/kubernetes/issues/67526) 
and [kubernetes #93111](https://github.com/kubernetes/kubernetes/issues/93111) 
where users expose the need to create a policy that allow a range of ports but some 
specific port, or also cases that a user wants to create a policy that allows 
the egress to other cluster to the NodePort range (eg 32000-32768) and in this case, 
the rule should be created specifying each port separately, as:

```
spec:
  egress:
  - ports:
    - protocol: TCP
      port: 32000
    - protocol: TCP
      port: 32001
    - protocol: TCP
      port: 32002
    - protocol: TCP
      port: 32003
[...]
    - protocol: TCP
      port: 32768
```

So for the user, actually:
* To allow a range of ports, each of them must be declared as an item from ``ports`` array
* To make an exception needs a declaration of all ports but the exception

Adding a new ``PortRange`` field inside the ``ports`` will allow a simpler 
creation of NetworkPolicy to the user.

### Goals

Add Range field to Ports in NetworkPolicy

### Non-Goals

TBD

> ## Proposal

Reviewer: is there a section for Alternate Proposal, where we can outline the proposal with "Exception" and why we considered it as not a viable option .. for the sake of completeness ?

Author: We can maybe add this to Non-Goals, like:

* Specifying Exceptions to the range -> An exception can at this time be specified as multiple ranges not contemplating the exception. Also specifying an exception at this time may lead to much more complex implementation, needing to interact through all the ports[].ranges to see if the exception is inside any of them, and may lead to some confusion.

Author: @andrewsykim any thoughts on this also?

Reviewer 2: (mark this conversation resolved?)

> In NetworkPolicy specification, inside ``NetworkPolicyPort`` object struct 

Reviewer: could you also add information on what to expect when both the `Port` and `Range` fields are set? do we allow them to be set simultaneously ? perhaps a clarification about it

Author: OK, so here I'm leaving as:

"If both ``Port`` and ``Range`` are specified, they are cumulative and no further
validation might occur. It's up to the CNI to summarize both of the fields in a 
single rule."

I don't know if this is our ideal scenario, but I can't neither see why we shouldn't do this way, after all we're specifying some ports to be allowed, are they in singular field (Port) or range field (Range).

Sounds good?

Reviewer: agree with you.. at least we clarify the usage in KEP so it is not misunderstood

(eg. 30000-32767) as I don't know which port is going to be allocated on the 
other side, but don't want to create a rule for each of them.

### Notes/Constraints/Caveats

*  The technology used by the CNI provider might not support port range in a 
trivial way. As an example, OpenFlow does not have a way to specify port range
and this might be a problem for OVS based CNIs as commented in 
[kubernetes #67526](https://github.com/kubernetes/kubernetes/issues/67526#issuecomment-415170435).
The same way, eBPF based CNIs will need to populate their maps in a different way.

### Risks and Mitigations

CNIs will need to support the new field in their controllers. For this case 
we'll try to make broader communication with the main CNIs so they can be aware
of the new field.

## Design Details

API changes to NetworkPolicy:
* Add a new struct called ``NetworkPolicyPortRange`` as the following:
```
type NetworkPolicyPortRange struct {
    From uint16
    To   uint16
}
```
* Add a new field ``spec.ingress|egress.ports.range`` that points to the
new struct:
```
// NetworkPolicyPort describes a port to allow traffic on
type NetworkPolicyPort struct {

>     Range PortRange

Author: changed here, good catch :)

  - Estimated increase in size: New struct inside the object with two fields of 
  16 bits each
  - Estimated amount of new objects: N/A

* **Will enabling / using this feature result in increasing time taken by any 
operations covered by [existing SLIs/SLOs]?**
  N/A

* **Will enabling / using this feature result in non-negligible increase of 
resource usage (CPU, RAM, disk, IO, ...) in any components?**
  It might get some increase of resource usage by the CNI while parsing the 
  new field.

### Troubleshooting

_This section must be completed when targeting beta graduation to a release._

* **How does this feature react if the API server and/or etcd is unavailable?**
  As this feature is mainly used by CNI providers, the reaction with API server 
  and/or etcd being unavailable will be the same as before.

* **What are other known failure modes?**
  N/A

* **What steps should be taken if SLOs are not being met to determine the problem?**
  N/A

## Implementation History
- 2020-10-08 Initial [KEP PR](https://github.com/kubernetes/enhancements/pull/2079)

## Drawbacks

*  The technology used by the CNI provider might not support port range in a 

> trivial way. As an example, OpenFlow does not have a way to specify port range

Reviewer: actually we can leverage the tcp_dst field in ovs in the port/mask format to specify range of ports and reduce the number of flows required. This is documented at https://www.openvswitch.org/support/dist-docs-2.5/ovs-ofctl.8.txt  | grep tcp_dst

Excerpt from the doc:
/*
              Bitwise  matches  on transport ports are rarely useful in isola‐
              tion, but a group of them can be used to reduce  the  number  of
              flows  required  to  match  on  a range of transport ports.  For
              example, suppose that the goal is to match TCP source ports 1000
              to  1999,  inclusive.   One way is to insert 1000 flows, each of
              which matches on a single source port.  Another way is  to  look
              at the binary representations of 1000 and 1999, as follows:
*/

specify a new ``Range`` field composed of the minimum and maximum ports 
inside the range.

If both ``Port`` and ``Range`` are specified, they are cumulative and no further
validation might occur. It's up to the CNI to summarize both of the fields in a 
single rule.

### User Stories (Optional)

#### Story 1

I have an application that communicates with NodePorts of a different cluster 
and I want to allow the egress of the traffic only the NodePort range 
(eg. 30000-32767) as I don't know which port is going to be allocated on the 
other side, but don't want to create a rule for each of them.

### Notes/Constraints/Caveats

*  The technology used by the CNI provider might not support port range in a 
trivial way as described in [#drawbacks]

### Risks and Mitigations

CNIs will need to support the new field in their controllers. For this case 
we'll try to make broader communication with the main CNIs so they can be aware
of the new field.

## Design Details

API changes to NetworkPolicy:
* Add a new struct called ``NetworkPolicyPortRange`` as the following:
```
type NetworkPolicyPortRange struct {
    From         uint16
    To           uint16

>     Exceptions:  []uint16

Reviewer 3: Can this field just be called `Except`?

Author: sounds good

}
```
* Add a new field ``spec.ingress|egress.ports.range`` that points to the
new struct:
```
// NetworkPolicyPort describes a port to allow traffic on

> type NetworkPolicyPort struct {

Reviewer 3: Can we add the whole struct here instead of just the range field?

Author: Added. BTW the documentation says:

```
// The port on the given protocol. This can either be a numerical or named port on
	// a pod. If this field is not provided, this matches all port names and numbers.
	// +optional
	Port *intstr.IntOrString
```

I'm worried here about if we provide ``Range`` but no ``Port`` it will still match all Ports, so I've changed the struct doc to:

```
// The port on the given protocol. This can either be a numerical or named 
  // port on a pod. If this field is not provided but a Range is 
  // provided, this field is ignored. Otherwise this matches all port names and
  // numbers.
```

This validation will probably need to be changed in the CNI? Or I need to check if this is something from the API

Reviewer 4: How about:

```
  // The port on the given protocol. This can either be a numerical or named 
  // port on a pod. If this field is not provided, this matches all port names and
  // numbers, whether an endPort is defined or not.
```

Or were you actually proposing something different?

Author: IMO it's good. We're exactly proposing in validations (I think we are!) that endPort can only exist if Port is defined.

Also I think this comment is old, when we were still considering using PortRange (and not EndPort), the new comment was:

```
// The port on the given protocol. This can either be a numerical or named port on
	// a pod. If this field is not provided, this matches all port names and numbers.
```

But I've changed as you've proposed, it's clearer

>     Range NetworkPolicyPortRange

Reviewer 3: Might be good to add the API docs here now.

Reviewer 3: Also I _think_ this has to be a pointer to allow for disabled fields.

Author: changed, no idea either

Author: Answering myself, yes they should be a pointer: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required

Author: but otoh, we have ``Number int32`` in ``ServiceBackendPort`` and this is probably because int32 have a default nil (which is 0)? Will try

}
```

### Validations
The range will need to be validated, with the following scenarios:
* If there's a ``From`` or a ``To`` field defined, the other one must be defined.
* ``From`` needs to be less than or equal to ``To``
* All the ports in the ``Exceptions`` array must be inside the defined range.
* ``Exception`` can only be defined if ``From`` and ``To`` are also defined.
* Because ``ports`` is a superset of all ports specified in ``port`` and 
``range``, if a port is specified in at least one of the fields it should be 
allowed.

### Test Plan

> TBD

Reviewer 3: This is probably needs to be filled in since the we changed the status to `implementable`. I think something like this would be efficient:

Unit tests:
* test API validation logic
* test API strategy to ensure disabled fields 

E2E tests:
*  Add e2e tests exercising only the API operations for port ranges. Data-path validation should be done by CNIs.

### Graduation Criteria

#### Alpha 
- Add a feature gated new field to NetworkPolicy
- Communicate CNI providers about the new field
- Add validation tests in API

#### Beta
- ``PortRanges`` has been supported for at least 1 minor release

> - At least one CNI provider implements the new field

Reviewer 3: Personally I think only requiring one implementation is too loose of a requirement for this field to be on by default. Not sure what the magic number is but should at least be a handful of CNIs?

Author: Yes, it was thinking more about having a PR to be reviewed other then a magic number.

I'm tempted to follow the same number proposed by Jay in #2113  and change to:

```
- The name selector has been supported for at least 1 minor release.
- Four commonly used NetworkPolicy (or CNI providers) implement the new field, 
with generally positive feedback on its usage.
- Feature Gate is enabled by Default.```

> <!-- toc -->

Reviewer 2: You seem to be missing a title now

Author: Added, thanks!

specific port, or also cases that a user wants to create a policy that allows 
the egress to other cluster to the NodePort range (eg 32000-32768) and in this case, 
the rule should be created specifying each port separately, as:

```
spec:
  egress:
  - ports:
    - protocol: TCP
      port: 32000
    - protocol: TCP
      port: 32001
    - protocol: TCP
      port: 32002
    - protocol: TCP
      port: 32003
[...]
    - protocol: TCP
      port: 32768
```

So for the user:
* To allow a range of ports, each of them must be declared as an item from 
``ports`` array
* To make an exception needs a declaration of all ports but the exception

Adding a new ``PortRange`` field inside the ``ports`` will allow a simpler 
creation of NetworkPolicy to the user.

### Goals

Add Range field to Ports in NetworkPolicy

### Non-Goals

N/A

## Proposal

In NetworkPolicy specification, inside ``NetworkPolicyPort`` object struct 
specify a new ``Range`` field composed of the minimum and maximum ports 
inside the range.

If both ``Port`` and ``Range`` are specified, they are cumulative and no further
validation might occur. It's up to the CNI to summarize both of the fields in a 
single rule.

> ### User Stories (Optional)

Reviewer 2: It's odd that you don't have any stories inspired by the two issues linked above.

The single existing user story seems like it could potentially be solved in other ways than allowing arbitrary port ranges.

In general, I feel like this feature is not an obvious must-have for NetworkPolicy, and so I feel like the KEP needs to make a strong case for it. And that means making a case that (a) there are use cases that fit within the design of NetworkPolicy (ie, application-focused policy, _not_ administrator-focused policy) which cannot easily be solved with the current NetworkPolicy features, but which could be solved if we added port ranges, _and_ (b) the port range solution does a better job of solving the use cases than any other obvious alternatives

Author: You're right :)

I've added new user stories, trying to make this more clarifying (following Winship's 101 User Story course from your Kep :P ) so let's see if with the new User Stories people get convinced this is necessary.

#### Story 1

I have an application that communicates with NodePorts of a different cluster 
and I want to allow the egress of the traffic only the NodePort range 
(eg. 30000-32767) as I don't know which port is going to be allocated on the 
other side, but don't want to create a rule for each of them.

### Notes/Constraints/Caveats

*  The technology used by the CNI provider might not support port range in a 
trivial way as described in [#drawbacks]

### Risks and Mitigations

CNIs will need to support the new field in their controllers. For this case 
we'll try to make broader communication with the main CNIs so they can be aware
of the new field.

## Design Details

API changes to NetworkPolicy:
* Add a new struct called ``NetworkPolicyPortRange`` as the following:
```
// NetworkPolicyPortRange describes the range of ports to be used in a 
// NetworkPolicyPort struct
type NetworkPolicyPortRange struct {
    // From defines the start of the port range
    From         uint16
    
    // To defines the end of the port range, being the end included within the
    // range
    To           uint16

    // Except defines all the exceptions in the port range

>     Except:      []uint16

Reviewer 2: wait, I thought exceptions were going away?

Author: @danwinship Last policy api meeting we've decided that the efforts to add exceptions again is too low compared to the benefits of adding it in the struct.

Basically, it's clearer when you define an exception inside a range instead of defining multiple ranges excluding an exception, like:

```
- from: 32000
  to: 32999
- from: 33001 # Excluding the 33000
  to: 34000
```
vs
```
- from: 32000
  to: 34000
  except:
  - 33001
```

So I've reinserted this in the struct, and removed this as a non goal :)

Reviewer 2: Well, ok, but what is the use case for exceptions then, such that they need to be easy to express? What is the developer use case for granting a pod access to 32000-34000 _except_ 33001?

There were strong objections to adding `except` to `ipBlock`, but we kept it because it's incredibly difficult to do the equivalent of

    - from: 32000
      to: 32999
    - from: 33001
      to: 34000

with CIDRs. if we didn't have `except` then in order to write

    - ipBlock:
        cidr: 192.168.0.0/24
        except:
          - 192.168.0.5

you'd have to do:

    - cidr: 192.168.0.128/25
    - cidr: 192.168.0.64/26
    - cidr: 192.168.0.32/27
    - cidr: 192.168.0.16/28
    - cidr: 192.168.0.8/29
    - cidr: 192.168.0.6/31
    - cidr: 192.168.0.4/32
    - cidr: 192.168.0.2/31
    - cidr: 192.168.0.1/32

which is both horrifying and incomprehensible (and probably also wrong; I didn't double-check it).

Author: I think this user story is still compliant, even if we've removed the exception, thinking that the user can create multiple ranges excluding the excepts, right?

>     Except:      []uint16

Reviewer 2: ```suggestion
    Except       []uint16
```

Author: Solved, thanks!

}
```
* Add a new field ``spec.ingress|egress.ports.range`` that points to the
new struct:
```
// NetworkPolicyPort describes a port or a range of ports to allow traffic on
type NetworkPolicyPort struct {
  // The protocol (TCP, UDP, or SCTP) which traffic must match. If not specified, this
  // field defaults to TCP.
  // +optional
  Protocol *api.Protocol
  
  // The port on the given protocol. This can either be a numerical or named 
  // port on a pod. If this field is not provided but a Range is 
  // provided, this field is ignored. Otherwise this matches all port names and
  // numbers.
  // +optional

>   Port *intstr.IntOrString

Reviewer 3: I'm still on the fence about this, but I feel that `Port` and `Range` under the same field feels awkward. Putting range semantics into it's own struct _might_ be cleaner, even if you have to specify the Protocol field twice. This way consumers don't have to reason about precedent between `port` and `range` under the same field, but can just take the union of `ports` and `portRanges`.  

```
ports:
  - protocol: TCP
    port: 80
  - protocol: TCP
    port: 443
portRanges:
  - protocol: TCP
    range:
      from: 30000
      to: 32000
```

Author: tbh I have no strong opinion about naming either, so I can move this to outside ports, making it an array and that's fine

Reviewer 3: Let's wait to see what Tim thinks :)

Author: Great, thank you for all the feedback Andrew :)

Reviewer 2: I agree `portRanges` is clearer.

Author: @andrewsykim @danwinship about @thockin proposal of just adding an EndPort into the existing ``NetworkPolicyPort`` sounds good and this conversation can be resolved?

  // A range of ports on a given protocol and the exceptions. If this field 
  // is not provided, this doesn't matches anything
  // +optional
  Range *NetworkPolicyPortRange
}
```

### Validations
The range will need to be validated, with the following scenarios:

> * If there's a ``From`` or a ``To`` field defined, the other one must be defined.

Reviewer 2: They are not optional fields, as currently defined

* ``From`` needs to be less than or equal to ``To``

> * All the ports in the ``Exceptions`` array must be inside the defined range.

Reviewer 3: This is probably a minor thing, but will we reject NetworkPolicys that have duplicate ports in `Except`?

Author: I don't think this is actually a big problem.

A similar question: what happens if I create two ``NetworkPolicyPort`` specifying the same port? I've tested here and the API accepts so I think we're good with repeated excepts as well

Author: Removed except, closing this conversation :)

> * All the ports in the ``Exceptions`` array must be inside the defined range.
> * ``Exception`` can only be defined if ``From`` and ``To`` are also defined.

Reviewer 2: ```suggestion
* All the ports in the ``Except`` array must be inside the defined range.
* ``Except`` can only be defined if ``From`` and ``To`` are also defined.
```

but again, `From` and `To` are currently required to always be defined

``range``, if a port is specified in at least one of the fields it should be 
allowed.
* If ``Range`` is defined but no ``Port`` is defined, the old behavior of matching
all should be changed.

### Test Plan

Unit tests:
* test API validation logic
* test API strategy to ensure disabled fields

E2E tests:
* Add e2e tests exercising only the API operations for port ranges. Data-path 
validation should be done by CNIs.

### Graduation Criteria

#### Alpha 
- Add a feature gated new field to NetworkPolicy
- Communicate CNI providers about the new field
- Add validation tests in API

#### Beta
- ``PortRanges`` has been supported for at least 1 minor release
- Four commonly used NetworkPolicy (or CNI providers) implement the new field, 
with generally positive feedback on its usage.
- Feature Gate is enabled by Default.

#### GA Graduation

- At least **four** NetworkPolicy providers (or CNI providers) support the ``PortRanges`` field
- ``PortRanges`` has been enabled by default for at least 1 minor release

### Upgrade / Downgrade Strategy

If upgraded no impact should happen as this is a new field.

If downgraded the CNI wont be able to look into the new field, as this does not 

> exists and network policies using this field will stop working.

Reviewer 2: They will stop working _correctly_ and start working incorrectly.

 N/A

* **Are there any missing metrics that would be useful to have to improve observability 
of this feature?**
 N/A

### Dependencies

* **Does this feature depend on any specific services running in the cluster?**
  No

### Scalability

_For alpha, this section is encouraged: reviewers should consider these questions
and attempt to answer them._

_For beta, this section is required: reviewers must answer these questions._

_For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field._

* **Will enabling / using this feature result in any new API calls?**
  TBD

* **Will enabling / using this feature result in introducing new API types?**
  No, unless the new ``NetworkPolicyPortRange`` is considered a new API type

* **Will enabling / using this feature result in any new calls to the cloud 
provider?**
  No

* **Will enabling / using this feature result in increasing size or count of 
the existing API objects?**

  - API type(s): NetworkPolicy / NetworkPolicyPorts
  - Estimated increase in size: New struct inside the object with two fields of 

>   16 bits each + 16 bits for each port in the ``Exceptions`` array

Reviewer 2: ```suggestion
  16 bits each + 16 bits for each port in the ``Except`` array
```

  - [Dependencies](#dependencies)
  - [Scalability](#scalability)
  - [Troubleshooting](#troubleshooting)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

## Release Signoff Checklist

Items marked with (R) are required *prior to targeting to a milestone / release*.

- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
- [ ] (R) KEP approvers have approved the KEP status as `implementable`
- [ ] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] (R) Graduation criteria is in place
- [ ] (R) Production readiness review completed
- [ ] Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

## Summary

> Today the ``ports``field in ingress and egress network policies is an array 

Reviewer 4: You only need a single backtick

and also to the range 49152-65535 without allowing any other ports.

### Notes/Constraints/Caveats

*  The technology used by the CNI provider might not support port range in a 
trivial way as described in [#drawbacks]

### Risks and Mitigations

CNIs will need to support the new field in their controllers. For this case 
we'll try to make broader communication with the main CNIs so they can be aware
of the new field.

## Design Details

API changes to NetworkPolicy:
* Add a new field called ``EndPort`` inside ``NetworkPolicyPort`` as the following:
```
// NetworkPolicyPort describes a port to allow traffic on
type NetworkPolicyPort struct {
	// The protocol (TCP, UDP, or SCTP) which traffic must match. If not specified, this
	// field defaults to TCP.
	// +optional
	Protocol   *v1.Protocol `json:"protocol,omitempty" protobuf:"bytes,1,opt,name=protocol,casttype=k8s.io/api/core/v1.Protocol"`

	// The port on the given protocol. This can either be a numerical or named port on
	// a pod. If this field is not provided, this matches all port names and numbers.
	// +optional
	Port       *intstr.IntOrString `json:"port,omitempty" protobuf:"bytes,2,opt,name=port"`

> 	// EndPort defines the end of the port range, being the end included within the

Reviewer 4: In terms of wording maybe:

`EndPort defines the last port included in the port range.`

and maybe an example?

Author: Ok, I've inserted  a 'simple' example. When you say example is in the code (that turns into docs) right? Don't know if this is right, will take a look into other parts of k/k code late tomorrow

	// range
	// +optional
	EndPort    int32 `json:"port,omitempty" protobuf:"bytes,2,opt,name=endPort"`
}
```

### Validations
The ``NetworkPolicyPort`` will need to be validated, with the following scenarios:

> * If an ``EndPort`` is specified a ``Port`` must also be specified

Reviewer 4: `EndPort` must be >= `Port` (note that == is allowed - is that OK?)

Author: great, forgot this case. IMO it's ok, == represents a single port. Don't know why someone would like to have a Port and an EndPort equal (compliance? etc) but seems reasonable to me

> * If ``Port`` is a string ``EndPort`` cannot be specified

Reviewer 4: I think this limitation is OK, but not strictly required

Reviewer: what would be the interpretation of 

```
Port: http
EndPort: 9000
```
?
maybe it is better to take a stand and not allow this?

Author: +1 not to allowing this. This can lead to a conversion confusion to CNIs. One concern I had was: does iptables supports named ports? Otherwise we could assume those ports are always converted to int by the CNI before inserting into the dataplane.

The answer is: yes, it accepts. Because we don't know how this is done in the wild (if there's some CNI creating rules with the named ports, and not the numbered ones) I think it's better also to say 'if you want a range, you can only use integer as the start of the range'

Reviewer 4: Whatever implements policy must convert named ports to numeric.

We *could* allow range here but it feels like a super corner case and let's just not.

### Test Plan

Unit tests:
* test API validation logic
* test API strategy to ensure disabled fields

E2E tests:
* Add e2e tests exercising only the API operations for port ranges. Data-path 
validation should be done by CNIs.

### Graduation Criteria

#### Alpha 
- Add a feature gated new field to NetworkPolicy
- Communicate CNI providers about the new field
- Add validation tests in API

#### Beta
- ``EndPort`` has been supported for at least 1 minor release
- Four commonly used NetworkPolicy (or CNI providers) implement the new field, 
with generally positive feedback on its usage.
- Feature Gate is enabled by Default.

#### GA Graduation

- At least **four** NetworkPolicy providers (or CNI providers) support the ``EndPort`` field
- ``EndPort`` has been enabled by default for at least 1 minor release

### Upgrade / Downgrade Strategy

If upgraded no impact should happen as this is a new field.

If downgraded the CNI wont be able to look into the new field, as this does not 
exists and network policies using this field will stop working correctly and

> start working incorrectly.

Reviewer 4: but they will fail-closed so it is acceptable

*  The technology used by the CNI provider might not support port range in a 
trivial way as described in [#drawbacks]

### Risks and Mitigations

CNIs will need to support the new field in their controllers. For this case 
we'll try to make broader communication with the main CNIs so they can be aware
of the new field.

## Design Details

API changes to NetworkPolicy:
* Add a new field called `EndPort` inside `NetworkPolicyPort` as the following:
```
// NetworkPolicyPort describes a port to allow traffic on
type NetworkPolicyPort struct {
	// The protocol (TCP, UDP, or SCTP) which traffic must match. If not specified, this
	// field defaults to TCP.
	// +optional
	Protocol   *v1.Protocol `json:"protocol,omitempty" protobuf:"bytes,1,opt,name=protocol,casttype=k8s.io/api/core/v1.Protocol"`

	// The port on the given protocol. This can either be a numerical or named 
  // port on a pod. If this field is not provided, this matches all port names and
  // numbers, whether an endPort is defined or not.
	// +optional
	Port       *intstr.IntOrString `json:"port,omitempty" protobuf:"bytes,2,opt,name=port"`

	// EndPort defines the last port included in the port range.
	// Example:

>   //    endPort: 12345

Reviewer 4: for example I meant something like `{ port: 8000, endport: 8002 } // ports 8000, 8001, 8002`

Reviewer 4: maybe an example is dumb here, don't sweat it. :)

undesired traffic.

There are a number of user issues like [kubernetes #67526](https://github.com/kubernetes/kubernetes/issues/67526) 
and [kubernetes #93111](https://github.com/kubernetes/kubernetes/issues/93111) 
where users expose the need to create a policy that allow a range of ports but some 
specific port, or also cases that a user wants to create a policy that allows 
the egress to other cluster to the NodePort range (eg 32000-32768) and in this case, 
the rule should be created specifying each port separately, as:

```
spec:
  egress:
  - ports:
    - protocol: TCP
      port: 32000
    - protocol: TCP
      port: 32001
    - protocol: TCP
      port: 32002
    - protocol: TCP
      port: 32003
[...]
    - protocol: TCP
      port: 32768
```

So for the user:
* To allow a range of ports, each of them must be declared as an item from 
``ports`` array
* To make an exception needs a declaration of all ports but the exception

Adding a new ``PortRange`` field inside the ``ports`` will allow a simpler 
creation of NetworkPolicy to the user.

### Goals

Add Port Range field in a NetworkPolicy

### Non-Goals

N/A

## Proposal

> In NetworkPolicy specification, inside ``NetworkPolicyIngressRule`` and 

Reviewer 4: If I read correctly this is not a compatible change.

If the user specifies `portRanges` and does not specify `ports` they hit this part of the spec:

```
type NetworkPolicyIngressRule struct {
    // List of ports which should be made accessible on the pods selected for this
    // rule. Each item in this list is combined using a logical OR. If this field is
    // empty or missing, this rule matches all ports (traffic not restricted by port).
    // If this field is present and contains at least one item, then this rule allows
    // traffic only if the traffic matches at least one port in the list.
    // +optional
    Ports []NetworkPolicyPort
```

Specifically "If this field is empty or missing, this rule matches all ports"

In other words, a back-rev implementation which does not understand the new field will fail-open, which is a no-no.

What if, instead, you added a "size" or "rangeSize" or "endPort" field to `NetworkPolicyPort` which converted the single port into a range.  That way, a back-rev implementation who did not know about the new field would allow access to/from the single "base" port"  and nothing else.  Still not the correct interpretation, but it fails closed.

I *think* that could work.  Could use extra brains on it.

Author: Same problem from namespaceName selector, right? What if the CNI does not know the new field, got it.

I see value here in having the endPort (don't think rangeSize is very intuitive, as you need to make some 'math' to calculate big ranges)..

Reviewer 5: Calico's datamodel supports port ranges by specifying the start and end port; this has worked well and I can't remember anyone having any issues grokking it; @thockin's idea sounds like a least-bad way to handle back compat.

Agree with @rikatz that rangeSize is not intuitive.

Could the struct be renamed to `NetworkPolicyPortRange` and still preserve back-compat (since the struct name doesn't actually feature in the JSON/YAML).

Reviewer: > Calico's datamodel supports port ranges by specifying the start and end port; this has worked well and I can't remember anyone having any issues grokking it; @thockin's idea sounds like a least-bad way to handle back compat.
> 
> Agree with @rikatz that rangeSize is not intuitive.
> 
> Could the struct be renamed to `NetworkPolicyPortRange` and still preserve back-compat (since the struct name doesn't actually feature in the JSON/YAML).

renaming the struct would probably break anyone referring to the type downstream?

Author: Yes, I guess. I'm afraid of someone referencing the struct name and breaking things. I think probably the best now is to specify the endPort (assuming the endPort is also part of the range)

Reviewer 5: > renaming the struct would probably break anyone referring to the type downstream?

Could leave a type alias in its place.  No big deal either way.

validation might occur. It's up to the CNI to summarize both of the fields in a 
single rule.

### User Stories

#### Story 1 - Opening communication to NodePorts of other cluster

I have an application that communicates with NodePorts of a different cluster 
and I want to allow the egress of the traffic only the NodePort range 
(eg. 30000-32767) as I don't know which port is going to be allocated on the 
other side, but don't want to create a rule for each of them.

#### Story 2 - Blocking the egress for not allowed insecure ports
As a developer, I need to create an application that scrapes informations from 
multiple sources, being those sources databases running in random ports, web 
applications and other sources. But the security policy of my company asks me 
to block communication with well known ports, like 111 and 445, so I need to create
a network policy that allows me to communicate with any port except those two and so
I can be compliant with the company's policy.

#### Story 3 - Containerized Passive FTP Server
As a Kubernetes User, I've received a demand from my boss to run our FTP server in an 
existing Kubernetes cluster, to support some of my legacy applications. 

> his FTP Server must be acessible from inside the cluster and outside the cluster, 

Reviewer 6: This FTP Server?

but I still need to keep the basic security policies from my company, that demands 
the existence of a default deny rule for all workloads and allowing only specific ports.

Because this FTP Server runs in PASV mode, I need to open the Network Policy to ports 21 
and also to the range 49152-65535 without allowing any other ports.

### Notes/Constraints/Caveats

> *  The technology used by the CNI provider might not support port range in a 

Reviewer 4: We should cross-check this with the biggest implementations to see if it is feasible.

Reviewer 5: FWIW, Calico has native support for port ranges (in iptables and BPF modes) so this would be straightforward for us.  iptables has native support for port ranges so any solution that uses that should be fine.

Reviewer: For OpenFlow based implementations, port ranges can be achieved using masks. At least in Antrea we can adapt to this change. I believe Cilium also natively supports range of ports in CliliumNetworkPolicies.

trivial way as described in [#drawbacks]

### Risks and Mitigations

CNIs will need to support the new field in their controllers. For this case 
we'll try to make broader communication with the main CNIs so they can be aware
of the new field.

## Design Details

API changes to NetworkPolicy:
* Add a new struct called ``NetworkPolicyPortRange`` as the following:
```
// NetworkPolicyPortRange describes the range of ports to be used in a 
// NetworkPolicyPort struct
type NetworkPolicyPortRange struct {
    // From defines the start of the port range
    From         uint16
    
    // To defines the end of the port range, being the end included within the
    // range
    To           uint16

    // Except defines all the exceptions in the port range
    +optional

>     Except       []uint16

Reviewer 6: Just wanted to clarify here if the expected behavior of the Except block, like in NetworkPolicy ipBlock, is fallthrough rather than explicit deny? See https://github.com/kubernetes/kubernetes/issues/83191 for more information. I'm mentioning this because if this is the case, then for the `User Story 2` you brought up in the previous section, it might be important to note this Except block will not create explicit port deny rule.

Author: Yeah, right :)

So by the other discussions, we've actually decided to drop the Exception instead of use it. This is going to bring too much confusion for the users and can be achieved by creating different port ranges (see https://kubernetes.slack.com/archives/C01CWSHQWQJ/p1605898496104400).

The thing is, if we decide to add the exception later this might be a breaking change because CNIs not aware of this may fail open a range that you insert an exception.

By now, IMO it's better that we keep this simple and drop the Exception. Sounds good?

Reviewer 6: Makes sense to me

>     Except       []uint16

Reviewer 4: I'm not super convinced we need Except.  CIDRs have to be powers of 2, which makes exceptions hard.  Port ranges have no such requirement, so any exception can be easily represented as 2 ranges.

Author: This was the original proposal, but we've discussed about the hardness of implementing this.

OTOH, by the last comment of adding the endPort field creating exceptions can be harder (unless Expections are also part of NetworkPolicyPort).

I don't have a strong opinion actually about the exception thing, so will leave for other folks to comment.

Reviewer 5: Except clause seems like overkill to me too.  It makes validation a bit more fiddly (exceptions must lie in the range, presumably) and of course, once you have an exception ports, why not exception ranges?  From an implementer's point of view, it makes it more fiddly but not by much.

- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

## Summary

<!--
This section is incredibly important for producing high-quality, user-focused
documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

> Today the ``ports``field in ingress and egress network policies is an array that needs a declaration of each single port to be contemplated. This KEP proposes to add a new field that allows a declaration of a port range, simplifying the creation of rules with multiple ports.

Reviewer 2: Please wrap lines to make it easier to review

## Motivation

<!--
This section is for explicitly listing the motivation, goals and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

NetworkPolicy object is a complex object, that allows a developer to specify what's the traffic behavior expected of the application and allow/deny undesired traffic.

> There are a number of user issues like [this](https://github.com/kubernetes/kubernetes/issues/67526) and [this](https://github.com/kubernetes/kubernetes/issues/93111) where users expose the need to create a policy that allow a range of ports but some specific port, or also cases that a user wants to create a policy that allows the egress to other cluster to the NodePort range (eg 32000-32768) and in this case, the rule should be created specifying each port separately, as:

Reviewer 2: Preferred KEP style is to use "reference" style links like the "`[experience reports]`" example in the comment just above, and refer to the issues explicitly (eg "`a number of user issues like [kubernetes #67526]`")

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->
Add PortRange field to Ports in NetworkPolicy

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

TBD
## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. The "Design Details" section below is for the real
nitty-gritty.
-->

In NetworkPolicy specification, inside ``NetworkPolicyPort`` object struct specify a new ``PortRange`` field:

```
// NetworkPolicyPort describes a port to allow traffic on
type NetworkPolicyPort struct {
    PortRange *string
}
```

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

> As the owner of a web application I want to allow the traffic to my Pods in different ports without having to create multiple network policies. As an example, my application is exposed in ports 80, 443 and 8443 and I want to group this in a single rule

Reviewer 2: You can do that now, you just need to specify

    - ports:
        - port: 80
        - port: 443
        - port: 8443

So why isn't that good enough for this user? (If your answer is "well this is just an example, in other cases you couldn't do this", then throw out this example and replace it with one of those better cases.)

(Also, I assume this means your proposal for `portRange` is going to allow comma-separated ports, but you don't say that anywhere.)

Author: You are right. I'm dropping this user story then. About the comma-separated, we've reached to api-machinery which said it's not a good idea as we would need to leave the validation of a string also to the API. https://kubernetes.slack.com/archives/CJUQN3E4T/p1603141881023900

#### Story 2

I have another application that communicates with NodePorts of a different cluster and I want to allow the egress of the traffic only the NodePort range (eg. 30000-32767) as I don't know which port is going to be allocated in the other side, but don't want to create a rule for each of them.

#### Story 3

> As the owner of an application, I want to allow the communication with a set of ports between 6000 and 7000 except port 6667.

Reviewer 2: This is not a user story, it's just a restatement of the proposed feature.

I can write out "As a user I want to give all of my money to @danwinship", but if I don't provide any more details/justification then that, then I wouldn't expect anyone else to believe that there are actually real users out there who need that feature. (But there totally are.)

Author: OK, the user story has been removed anyway as it doesn't make sense now

> ### Notes/Constraints/Caveats (Optional)

Reviewer 2: As mentioned in one of the issues, this feature would be difficult/inefficient to implement in Open vSwitch-based network plugins (eg, openshift-sdn, ovn-kubernetes, Antrea), since OpenFlow does not have any way to specify port ranges. There may also be other network types that have difficulty implementing this as well.

Author: @danwinship I couldn't find the comment in the previous issues, but anyway I'm taking a look into OVS stuff to check the constraint (I believe you, just using this as an excuse to study it better).

If you have the comment can you please provide it so I can feed the document? 

Anyway I'm inserting this as a general caveat, it depends if the technology used by the CNI provider supports multiport. It came to my mind now that this might be a problem as well to eBPF based CNIs that would need to change how they populate the maps used by programs as well, so maybe it would be a need of the CNI to loop through all the ports and generate a rule for each of them.

@aojea also pointed into a [part of code](https://github.com/kubernetes/kubernetes/blob/97a13a0/pkg/security/podsecuritypolicy/provider.go#L421) that converts an existing ``HostPortRange`` field to comma-separated but it seems to me it's mostly related to PSP so wondering how this also could be re-used.

Anyway, I'll point that in the doc.

Author: Found it, never mind

Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

CNIs will need to support the new field in their controllers. For this case we'll try to make broader communication with the main CNIs so they can be aware of the new field.

The field will need some syntax validation either, to not allow the mix of named and non-named ports inside a range (like http-9000), and also to allow only numbered ports inside uint16 range.

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

API changes to NetworkPolicy:

> * Add a new field ``spec.ingress|egress.ports.PortRange

Reviewer 3: Wondering if it makes sense to have `portRange` under `ports` or should `ports` and `portRanges` be separate fields under `ingress` and `egress`? What happens if `portRanges` conflicts with `port`?

Author: I think it's up to the CNI how it's going to deal with the conflicts? Or should we validate this in the API?

 But I have no opposition to move portRanges to outside ports, whatever is cleaner/better

Reviewer: if we move port under portRange(or maybe some new name, so having port under portRange doesn't sound awkward) and validate at API level that only one of those fields is set, we can eliminate potential conflicts

Author: do we want to eliminate the conflict, or consider only one field? 

Or maybe convert a .port entry to a .portRange.port as proposed in the comment below

Also, about the naming thing, maybe .ports.Range can be a better naming?

Reviewer 3: I think it would be useful to see all the example yaml for all the options, maybe in a separate google doc for now?

> * On field creation/change a validation should happen to verify if the format is compliant with ``nnnnn,nnnn1-nnnn5,http`` where nnnnn is a number between 0 and 65535 and named ports aren't mixed with numbered ports

Reviewer 3: I don't think we've had much discussion yet on what the API for this should look like, maybe something to discuss on Monday's call?

I'm a bit opposed to putting the range semantic in the field value like `nnnn1-nnnn5`, I'm wondering if instead we might want explicit fields for `to`, `from` and `except`, something like:
```
- ingress:
  - portRange:
      from: 30000
      to: 32000
      except:
          - 31000
          - 31001
```  

Wondering how exceptions are expressed with ``nnnnn,nnnn1-nnnn5,http``, maybe I missed that.

Author: Yeah, I remember there was a discussion somewhere about the syntax, but because I couldn't find I've wrote whatever came to my mind expected you folks to jump in and discuss with me the best syntax :)

or by redefining what graduation means.

In general we try to use the same stages (alpha, beta, GA), regardless of how the
functionality is accessed.

[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/

Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
-->
#### Alpha 
- Add a feature gated new field to NetworkPolicy
- Communicate CNI providers about the new field
- Add validation tests in API

#### Beta
- ``PortRanges`` has been supported for at least 1 minor release
- At least one CNI provider implements the new field
- Feature Gate is enabled by Default

#### GA Graduation

- At least two CNIs support the ``PortRanges`` field
- ``PortRanges`` has been enabled by default for at least 1 minor release

**Note:** Generally we also wait at least two releases between beta and
GA/stable, because there's no opportunity for user feedback, or even bug reports,
in back-to-back releases.

> #### Removing a Deprecated Flag

Reviewer 2: You should get rid of as much template text as you can to make it easier to see the actual KEP text. eg, just delete any sections (like this one about command-line flags) which are irrelevant to your KEP.
